Fail start of non-data node if node has data

henningandersen opened the pull request:
Check that nodes started with node.data=false cannot start if they have
shard data to avoid (old) indexes being resurrected into the cluster in red status.

Issue #27073

Only fixes node.data=false part of issue, not coordinating only.

I were not sure if this should go into docs too?

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -309,6 +310,11 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
                 ensureAtomicMoveSupported(nodePaths);
             }
+
+            if (!DiscoveryNode.isDataNode(settings)) {
```
we prefer the `X == false` convention over `!X` (see other code) in order to make the negation more visually prominent.

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
-     *
-     * @param indexSettings settings for the index
-     */
+    private void ensureNoIndexData(final NodePath[] nodePaths) throws IOException {
```
This should be called `ensureNoShardData`, as index data has less clear of a definition.

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
-     * @param indexSettings settings for the index
-     */
+    private void ensureNoIndexData(final NodePath[] nodePaths) throws IOException {
+        List<Path> badDataPaths = new ArrayList<>();
```
maybe `dataPathsWithShards`?

henningandersen replied:
Ended with shardDataPaths, since it contains the path to the shard data.

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
+            && path.getFileName().toString().chars().allMatch(Character::isDigit);
+    }
+
+        /**
```
the indentation looks off here.

henningandersen replied:
Thanks for catching this.

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
+        }
+
+        if (!badDataPaths.isEmpty()) {
+            throw new IllegalArgumentException("Node is started with "
```
`IllegalStateException` might be a better fit

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
+        if (!badDataPaths.isEmpty()) {
+            throw new IllegalArgumentException("Node is started with "
+                + Node.NODE_DATA_SETTING.getKey()
+                + "=false, but found shard data: "
```
maybe `"=false, but has shard data: "`

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java:
```diff
@@ -470,6 +471,72 @@ public void testExistingTempFiles() throws IOException {
+                .put(settings)
+                .put(Node.NODE_DATA_SETTING.getKey(), false).build();
+            newNodeEnvironment(noDataSettings).close();
+            fail("Must fail creating NodeEnvironment on a data path that has shard data if node.data=false");
```
you can use `expectThrows` which allows writing this in a more convenient way

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java:
```diff
@@ -470,6 +471,72 @@ public void testExistingTempFiles() throws IOException {
+        newNodeEnvironment(settings).close();
+    }
+
+    private String[] pathsToStrings(Path[] paths) {
```
this method is not used anywhere AFAICS

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java:
```diff
@@ -470,6 +471,72 @@ public void testExistingTempFiles() throws IOException {
 
+    public void testFailOnDataForNonDataNode() throws IOException {
+        Settings settings = buildEnvSettings(Settings.EMPTY);
+        final NodeEnvironment env = newNodeEnvironment(settings);
```
`buildEnvSettings` is called as part of `newNodeEnvironment`. Just calling `newNodeEnvironment()` here is sufficient.

henningandersen replied:
The reason I explicitly build settings here is to be able to reuse the path.node setting for the next newNodeEnvironment call. Is there a better way?

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java:
```diff
@@ -470,6 +471,72 @@ public void testExistingTempFiles() throws IOException {
+        newNodeEnvironment(noDataSettings).close();
+    }
+
+    public void testSucceedOnDataForDataNode() throws IOException {
```
maybe merge this test into `testFailOnDataForNonDataNode` above, calling it `test EnsureNoShardData`

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java:
```diff
@@ -470,6 +471,72 @@ public void testExistingTempFiles() throws IOException {
+
+        Index badIndex = new Index("bad", "badUUID");
+        for (Path path : env.indexPaths(badIndex)) {
+            Files.createDirectories(path.resolve("0"));
```
randomize the number here, e.g. using `randomInt(10)`

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java:
```diff
@@ -0,0 +1,76 @@
+        client().prepareIndex("test", "type1", "1").setSource("field1", "value1").setRefreshPolicy(IMMEDIATE).get();
+
+        logger.info("--> waiting for green status");
+        ensureGreen();
```
this is not necessary, prepareCreate already waits for all primaries to become active (= started), and there are no replicas in the test.

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java:
```diff
@@ -0,0 +1,76 @@
+        for (int i = 0; i < 10; i++) {
+            assertHitCount(client().prepareSearch().setQuery(matchAllQuery()).get(), 1L);
+        }
+        assertThat(client().prepareGet("test", "type1", "1").execute().actionGet().isExists(), equalTo(true));
```
why do a search and get here? we assume that these work (tested by other tests...)

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java:
```diff
@@ -0,0 +1,76 @@
+        assertThat(client().prepareGet("test", "type1", "1").execute().actionGet().isExists(), equalTo(true));
+
+        logger.info("--> restarting the node with node.data=true");
+        internalCluster().restartNode(node_1, new InternalTestCluster.RestartCallback());
```
just `restartRandomDataNode`

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java:
```diff
@@ -0,0 +1,76 @@
+        internalCluster().restartNode(node_1, new InternalTestCluster.RestartCallback());
+
+        logger.info("--> waiting for green status");
+        ensureGreen();
```
why is necessary to wait for green here?

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java:
```diff
@@ -0,0 +1,76 @@
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class NodeEnvironmentIT extends ESIntegTestCase {
+    public void testStartFailureOnDataForNonDataNode() throws Exception {
+        logger.info("--> starting two nodes");
```
why have two nodes in this test?

ywelsch commented on server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java:
```diff
@@ -0,0 +1,76 @@
+            });
+            fail("Starting a node with node.data=false that has index/shard data must fail");
+        } catch (IllegalArgumentException e) {
+            // EXPECTED
```
also assert something on the exception message?

henningandersen replied:
I ended up only asserting that the message contains the index UUID, would be nice to assert the entire path if possible, but I will need help if you think that is worth pursuing?

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
+            throw new IllegalArgumentException("Node is started with "
+                + Node.NODE_DATA_SETTING.getKey()
+                + "=false, but found shard data: "
+                + badDataPaths);
```
Is this the absolute path that we're outputting here?

henningandersen replied:
This would be the relative path if path.node is relative. Do you prefer the absolute path?

henningandersen replied:
Changed it to absolute path.

henningandersen commented:
Thanks for your comments @ywelsch, I have addressed mostt of them in the latest commit and added comments to the rest. Will you please take another look?

henningandersen commented:
@elasticmachine run gradle build tests 1

henningandersen commented:
@elasticmachine run gradle build tests 2

henningandersen commented:
@elasticmachine run gradle build tests 1

henningandersen commented:
@elasticmachine run gradle build tests 1

henningandersen commented:
@elasticmachine run gradle build tests 1

henningandersen commented:
@elasticmachine run gradle build tests 1

henningandersen commented:
@elasticmachine run gradle build tests 2
